manager/scheduler/constraint.go

```diff
@@ -44,29 +49,45 @@ func (f *ConstraintFilter) Check(n *NodeInfo) bool {
 			if !constraint.Match(n.Description.Hostname) {
 				return false
 			}
-		default:
-			// default is node label in form like 'node.labels.key==value'
-			// if it is not well formed, always fails it
-			if !strings.HasPrefix(constraint.Key, "node.labels.") {
+		case "node.role":
+			if !constraint.Match(n.Spec.Role.String()) {
 				return false
 			}
-			// if the node doesn't have any label,
-			// it's equivalent to match an empty value.
-			// that is, 'node.labels.key!=value' should pass and
-			// 'node.labels.key==value' should fail
-			if n.Spec.Annotations.Labels == nil {
-				if !constraint.Match("") {
+		default:
+			// default is label constraint in form like 'node.labels.key==value'
+			// or 'engine.labels.key!=value'
+			if strings.HasPrefix(constraint.Key, nodeLabelPrefix) {
```

Comment:
Make the `switch` statement a simple `switch {`, then this can be one of the case statements.